### PR TITLE
Update folder roles if they've been changed

### PR DESCRIPTION
**Pull request by hallamoore**
Fixes [N1/#1457](https://github.com/nylas/N1/issues/1457)

Summary:
We added a guessing mechanism to figure out which folders should be used for standard roles, but these guessed roles are only saved if it's the initial creation of a folder. Existing users when we implemented this have not seen any improvement in folder roles, even though the guessing mechanism seems to be returning the proper roles. This diff makes sure the roles get updated properly during sync.

Test Plan: Tested locally on a single IMAP account

Reviewers: khamidou

### CODE REVIEW

File: inbox/mailsync/backends/imap/monitor.py [hunk 1]
```diff
@@ -103,8 +102,9 @@ def save_folder_names(self, db_session, raw_folders):
 
         # Create new folders
         for raw_folder in raw_folders:
-            Folder.find_or_create(db_session, account, raw_folder.display_name,
-                                  raw_folder.role)
+            folder = Folder.find_or_create(db_session, account, raw_folder.display_name,
+                                           raw_folder.role)
+            Folder.update_role(folder, raw_folder.role)
```
Comment (bengotow) [inline inbox/mailsync/backends/imap/monitor.py]: Hmm - does this need to update the Category associated with the Folder as well? Updating only the folder might not have an impact on the API response to /folders or /labels.

File: inbox/models/folder.py [hunk 2]
```diff
@@ -76,10 +76,17 @@ def find_or_create(cls, session, account, name, role=None):
                      .format(name, account.id))
             raise
 
-        if not obj.canonical_name:
+        return obj
+
+
+    @classmethod
+    def update_role(cls, obj, role=''):
```
Comment (bengotow) [inline inbox/models/folder.py]: As long as we have both folder and category in the codebase, I think the best way to keep the two in sync might be to override the setters. That way, people don't have to know `update_role` exists, and it's not possible for them to accidentally assign the role in a way that makes things fall out of sync.

So for example, we could say:

```
    _role = Column(String(MAX_INDEXABLE_LENGTH), nullable=False,
                            default='', name="canonical_name")

    @property
    def role(self):
        return self._role

    @role.setter
    def role(self, value):
        self._role = value
        self.category.name = value
```

File: inbox/models/folder.py [hunk 3]
```diff
@@ -37,8 +37,18 @@ class Folder(MailSyncBase, UpdatedAtMixin, DeletedAtMixin):
...
-                            default='')
+    _canonical_name = Column(String(MAX_INDEXABLE_LENGTH), nullable=False,
+                             default='', name="canonical_name")
+
+    @property
+    def canonical_name(self):
+        return self._canonical_name
+
+    @canonical_name.setter
+    def canonical_name(self, value):
```
Comment (bengotow) [inline inbox/models/folder.py]: I think the ideal way to do the default value would be `value=''` in the function declaration!

**Comment by bengotow** [comment]
This LGTM—I definitely like this approach to binding the two together. I'll defer to @khamidou for a final LGTM though!

File: inbox/models/folder.py [hunk 4]
```diff
@@ -37,8 +37,18 @@ class Folder(MailSyncBase, UpdatedAtMixin, DeletedAtMixin):
...
+        value = value or ''
+        self._canonical_name = value
+        self.category.name = value
```
Comment (mhahnenberg) [inline inbox/models/folder.py]: I believe this broke some of the sync-engine tests because `self.category` can be None. Working on a fix...

**Pull request merged**